### PR TITLE
fix todo item in pkg/registry/rbac/validation/policy_comparator.go

### LINKED ISSUE #61703: kubectl auth reconcile output is too long
**Is this a BUG REPORT or FEATURE REQUEST?**:

> Uncomment only one, leave it on its own line: 
>
/kind bug
> /kind feature

**What happened**:
The object output is far beyond human-readable in `kubectl auth reconcile` command.
For example:
1. Create a role by xiaomi-role.yaml
xiaomi-role.yaml:
```
  kind: Role
  apiVersion: rbac.authorization.k8s.io/v1
  metadata:
    name: xiaomi
  rules:
  - apiGroups: ["extensions"]
    resources: ["deployments"]
    verbs: ["get", "list", "watch", "create", "update", "patch", "delete"] 
```
2. Here's another yaml file xiaomi-modify,yaml
```
  kind: Role
  apiVersion: rbac.authorization.k8s.io/v1
  metadata:
    name: xiaomi
  rules:
  - apiGroups: ["", "extensions", "apps"]
    resources: ["deployments", "replicasets", "pods"]
    verbs: ["get", "list", "watch", "create", "update", "patch", "delete"] 
```
3. execute `kubectl auth reconcile -f xiaomi-modify.yaml --output=yaml
`
Output of this command is 600+ lines, which is obvious too long to use.

**What you expected to happen**:
Squash the output
**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

### PR DESCRIPTION
**Pull request by WanLinghao**
**What this PR does / why we need it**:
In pkg/registry/rbac/validation/policy_comparator.go exists a todo item.
https://github.com/kubernetes/kubernetes/blob/3a2fe7b8f93e39c93b1f5e1c92f74c72c6671fc4/pkg/registry/rbac/validation/policy_comparator.go#L31
This todo item may cause a problem in` kubectl auth reconcile` command.
After this patch ,the output of this command will be squashed.
Example in
https://github.com/kubernetes/kubernetes/issues/61703
will be like:
```
{
    "kind": "Role",
    "apiVersion": "rbac.authorization.k8s.io/v1",
    "metadata": {
        "name": "xiaomi",
        "namespace": "office",
        "selfLink": "/apis/rbac.authorization.k8s.io/v1/namespaces/office/roles/xiaomi",
        "uid": "f92f8911-30e2-11e8-a88f-901b0e3ac8bc",
        "resourceVersion": "333",
        "creationTimestamp": "2018-03-26T10:46:43Z"
    },
    "rules": [
        {
            "verbs": [
                "get",
                "list",
                "watch",
                "create",
                "update",
                "patch",
                "delete"
            ],
            "apiGroups": [
                "extensions"
            ],
            "resources": [
                "deployments"
            ]
        },
        {
            "verbs": [
                "create",
                "delete",
                "get",
                "list",
                "patch",
                "update",
                "watch"
            ],
            "apiGroups": [
                "",
                "apps",
                "extensions"
            ],
            "resources": [
                "replicasets"
            ]
        },
        {
            "verbs": [
                "create",
                "delete",
                "get",
                "list",
                "patch",
                "update",
                "watch"
            ],
            "apiGroups": [
                "",
                "apps",
                "extensions"
            ],
            "resources": [
                "pods"
            ]
        },
        {
            "verbs": [
                "create",
                "delete",
                "get",
                "list",
                "patch",
                "update",
                "watch"
            ],
            "apiGroups": [
                "",
                "apps"
            ],
            "resources": [
                "deployments"
            ]
        }
    ]
}
```

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #61703

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by WanLinghao** [comment]
@liggitt the function you said just do a simple collapse, it collapses Verb field.
And also it will not collapse nonResourceURLs kind rules.
My patch collapses Verbs, APIGroups, Resources. nonResourceURLs will be collapsed too.
For example, we got two yaml files here:
old-clusterrole.yaml:
```
apiVersion: rbac.authorization.k8s.io/v1beta1
kind: ClusterRole
metadata:
  name: prometheus
rules:
- apiGroups: [""]
  resources:
  - nodes
  - services
  verbs: ["get", "list", "watch"]
- apiGroups: [""]
  resources:
  - configmaps
  verbs: ["get"]
- nonResourceURLs: ["/metrics"]
  verbs: ["get"]

```
new-clusterrole.yaml:
```
apiVersion: rbac.authorization.k8s.io/v1beta1
kind: ClusterRole
metadata:
  name: prometheus
rules:
- apiGroups: ["","extensions","v1"]
  resources:
  - nodes
  - services
  - endpoints
  - pods
  verbs: ["get", "list", "watch"]
- apiGroups: ["","gakki"]
  resources:
  - configmaps
  - jugg
  verbs: ["get","masami"]
- nonResourceURLs: ["/metrics","/*"]
  verbs: ["get","result","watch"]
```
Execute command as follows:
```
kubectl create -f old-clusterrole.yaml
kubectl auth reconcile -f new-clusterrole.yaml --output=yaml
```
And if use old function to complete squash, the output would be:
```
- nonResourceURLs:
  - /metrics
  verbs:
  - result
- nonResourceURLs:
  - /metrics
  verbs:
  - watch
- nonResourceURLs:
  - /*
  verbs:
  - get
- nonResourceURLs:
  - /*
  verbs:
  - result
- nonResourceURLs:
  - /*
  verbs:
  - watch
- apiGroups:
  - extensions
  resources:
  - pods
  verbs:
  - get
  - list
  - watch
- apiGroups:
  - v1
  resources:
  - nodes
  verbs:
  - get
  - list
  - watch
- apiGroups:
  - v1
  resources:
  - services
  verbs:
  - get
  - list
  - watch
- apiGroups:
  - v1
  resources:
  - endpoints
  verbs:
  - get
  - list
  - watch
- apiGroups:
  - ""
  resources:
  - configmaps
  verbs:
  - masami
- apiGroups:
  - extensions
  resources:
  - nodes
  verbs:
  - get
  - list
  - watch
- apiGroups:
  - v1
  resources:
  - pods
  verbs:
  - get
  - list
  - watch
- apiGroups:
  - ""
  resources:
  - endpoints
  verbs:
  - get
  - list
  - watch
- apiGroups:
  - extensions
  resources:
  - services
  verbs:
  - get
  - list
  - watch
- apiGroups:
  - ""
  resources:
  - jugg
  verbs:
  - get
  - masami
- apiGroups:
  - ""
  resources:
  - pods
  verbs:
  - get
  - list
  - watch
- apiGroups:
  - extensions
  resources:
  - endpoints
  verbs:
  - get
  - list
  - watch
- apiGroups:
  - gakki
  resources:
  - configmaps
  verbs:
  - get
  - masami
- apiGroups:
  - gakki
  resources:
  - jugg
  verbs:
  - get
  - masami
```

```
For my function, the output is much simple:
- apiGroups:
  - extensions
  - v1
  resources:
  - nodes
  - services
  verbs:
  - get
  - list
  - watch
- apiGroups:
  - ""
  - extensions
  - v1
  resources:
  - endpoints
  - pods
  verbs:
  - get
  - list
  - watch
- apiGroups:
  - gakki
  resources:
  - configmaps
  verbs:
  - get
  - masami
- apiGroups:
  - ""
  - gakki
  resources:
  - jugg
  verbs:
  - get
  - masami
- apiGroups:
  - ""
  resources:
  - configmaps
  verbs:
  - masami
- nonResourceURLs:
  - /*
  - /metrics
  verbs:
  - watch
  - result
- nonResourceURLs:
  - /*
  verbs:
  - get
```

In general, the function you mentioned complete simple collapse.
My patch improve this job. 
In most cases, output of this function is minimum version a  rule group can be collapsed

File: pkg/registry/rbac/validation/policy_comparator.go [hunk 1]
```diff
@@ -171,3 +176,176 @@ func ruleCovers(ownerRule, subRule rbac.PolicyRule) bool {
...
+
+	shouldJump = func(rule rbac.PolicyRule) bool {
+		return len(rule.Verbs) != 1
+	}
+	c.rules = collapse(c.rules, 8, 4, shouldJump)
+
+	return c
+}
+
+func collapse(rules []rbac.PolicyRule, keyFilter, appendFilter int, shouldJump func(rule rbac.PolicyRule) bool) []rbac.PolicyRule {
```
Comment (liggitt) [inline pkg/registry/rbac/validation/policy_comparator.go]: this approach is pretty complex to follow. for something so fundamental to the RBAC system, it needs to be much more straightforward. something like https://github.com/liggitt/audit2rbac/blob/master/pkg/util.go#L44 might be an easier approach to follow

File: pkg/registry/rbac/validation/policy_comparator.go [hunk 2]
```diff
@@ -171,3 +176,176 @@ func ruleCovers(ownerRule, subRule rbac.PolicyRule) bool {
...
+		}
+	}
+
+	for _, rule := range pool {
+		compacted = append(compacted, *rule)
+	}
+	return compacted
+}
+
+func generatorKey(rule rbac.PolicyRule, filter int) string {
```
Comment (liggitt) [inline pkg/registry/rbac/validation/policy_comparator.go]: this assumes no fields will be added to the PolicyRule struct and ignores them when determining a match if new fields are added

Comment (WanLinghao) [reply pkg/registry/rbac/validation/policy_comparator.go]: @liggitt I aggree with you.
If our object is collapsing a group of rules into minimum version, https://github.com/liggitt/audit2rbac/blob/master/pkg/util.go#L44 is not a mature solution. It still needs improvements.
But if what we need here is just a usable collapsed groups, I think your solution is  close to  be mature.
So , please inform me which version do we really need here.
And do you mind if I give a example code which follows your idea.
If not, I will work on this.
Thanks

**Comment by WanLinghao** [comment]
@liggitt  WDYT, or we can simply invoke the existed function you mentioned to do a simple collapse

**Comment by WanLinghao** [comment]
@liggitt I have updated this patch, now we just simply use the existed collapse function to do a simple collapse. Although simple, I think it could collapse the output into human-readable state in most cases.
PTAL, thanks

**Pull request closed without merging**